### PR TITLE
Remove `rejector` from platformvm block `executor`

### PR DESCRIPTION
**Pull request by dhrubabasu**
## Why this should be merged

Adds complexity for little gain IMO.

## How this works

pretty straightforward

## How this was tested

CI

### CODE REVIEW

File: vms/platformvm/block/executor/block.go [hunk 1]
```diff
@@ -24,7 +24,9 @@ var (
 // Exported for testing in platformvm package.
 type Block struct {
 	block.Block
-	manager *manager
+	manager                  *manager
+	rejected                 bool
```
Comment (StephenButtolph) [inline vms/platformvm/block/executor/block.go:28]: Why are we adding this `rejected` bool?

Comment (dhrubabasu) [reply vms/platformvm/block/executor/block.go:28]: To report it as `choices.Rejected` in `Status()`

**Review by abi87 — CHANGES_REQUESTED** [review]
I don't think removing the rejector lowers the complexity. In fact I think it increases it.
The idea of acceptor, rejector and verifier is that these three block actions are fully contained into separate objects. You don't need to look anywhere else to figure out how  verify/accept/reject are carried out.
I reckon the rejector is not doing much but I think ditching it in favour of some fields in the block actually increases the complexity, as it forces me to inspect the whole block and figure out by myself what pieces are relevant to rejection and what not.
If this is an attempt at a more general refactoring and I lack some context please let me know. Otherwise I really don't see reasons to carry out this change.

**Comment by dhrubabasu** [comment]
> I don't think removing the rejector lowers the complexity. In fact I think it increases it. The idea of acceptor, rejector and verifier is that these three block actions are fully contained into separate objects. You don't need to look anywhere else to figure out how verify/accept/reject are carried out. I reckon the rejector is not doing much but I think ditching it in favour of some fields in the block actually increases the complexity, as it forces me to inspect the whole block and figure out by myself what pieces are relevant to rejection and what not. If this is an attempt at a more general refactoring and I lack some context please let me know. Otherwise I really don't see reasons to carry out this change.

We've accrued a lot of tech debt in the platformvm and it's very confusing to understand what's going on. For this change in particular, `Block` consumers only call `Block.Reject()` which internally calls the `rejector` which in turn executes the same logic for every block type. Removing `rejector` and in-lining the logic directly into `Block` decreases the number of files you have to look at to realize all blocks execute the same logic when being rejected. This aligns with the AVM implementation: https://github.com/ava-labs/avalanchego/blob/v1.10.15/vms/avm/block/executor/block.go#L264-L295.

**Comment by abi87** [comment]
> > I don't think removing the rejector lowers the complexity. In fact I think it increases it. The idea of acceptor, rejector and verifier is that these three block actions are fully contained into separate objects. You don't need to look anywhere else to figure out how verify/accept/reject are carried out. I reckon the rejector is not doing much but I think ditching it in favour of some fields in the block actually increases the complexity, as it forces me to inspect the whole block and figure out by myself what pieces are relevant to rejection and what not. If this is an attempt at a more general refactoring and I lack some context please let me know. Otherwise I really don't see reasons to carry out this change.
> 
> We've accrued a lot of tech debt in the platformvm and it's very confusing to understand what's going on. For this change in particular, `Block` consumers only call `Block.Reject()` which internally calls the `rejector` which in turn executes the same logic for every block type. Removing `rejector` and in-lining the logic directly into `Block` decreases the number of files you have to look at to realize all blocks execute the same logic when being rejected. This aligns with the AVM implementation: https://github.com/ava-labs/avalanchego/blob/v1.10.15/vms/avm/block/executor/block.go#L264-L295. Happy to discuss offline but I don't think the extra file hop buys us much here

I agree that the P-chain has some technical debt, but I believe it is mostly around state management rather than this block manager entities. I am happy to signal interesting spots to cleanup if you are interested.
While the X-chain has a single block type, the P-chain has multiple ones; we have adopted a visitor pattern to handle the specificities of verifying/accepting different kind of blocks.
We use the visitor in many places of the P-chain codebases and this consistency I believe contributes to lower complexity. The mempool is another instance where, for the time being, we do use the visitor even if handling of stakerTxs and standartTxs doesn't differ that much.
I really don't see how special casing rejection and handling it in Blocks make things simpler.
Right now, if we need to understand verify/accept/reject logic, we go to each struct and all is there; with this proposal we'd do that for verify/accept but not for reject.
The extra file hop buys us a consistent way to handle business logic, that I believe we should preserv. At least until we only have a single block type to handle in the P-chain.
Happy to chat about this

File: vms/platformvm/block/executor/block.go (hunk 1 above)
Comment (danlaine) [reply vms/platformvm/block/executor/block.go:28]: Is this a behavior change from before? Previously, we didn't store a block being `rejected` but now we do, and report it in `Block.Status`

Comment (dhrubabasu) [reply vms/platformvm/block/executor/block.go:28]: Yes, this is a change. I can remove it but thought it'd be good to match the AVM

**Pull request closed without merging**